database.go

```diff
@@ -737,3 +740,203 @@ func marshalTags(tags map[string]string) []byte {
+	// Find measurement by name.
+	m := dbi.db.measurements[name]
+	if m == nil {
+		return 0, influxql.Unknown
+	}
+
+	// Find field by name.
+	f := m.FieldByName(field)
+	if f == nil {
+		return 0, influxql.Unknown
+	}
+
+	return f.ID, f.Type
+}
+
+// CreateIterator returns an iterator given a series data id, field id, & field data type.
+func (dbi *dbi) CreateIterator(seriesID uint32, fieldID uint8, typ influxql.DataType, min, max time.Time, interval time.Duration) influxql.Iterator {
+	// TODO: Add retention policy to the arguments.
+
+	// Create an iterator to hold the transaction and series ids.
+	itr := &iterator{
+		seriesID: seriesID,
+		fieldID:  fieldID,
+		typ:      typ,
+		imin:     -1,
+		interval: int64(interval),
+	}
+	if !min.IsZero() {
+		itr.min = min.UnixNano()
+	}
+	if !max.IsZero() {
+		itr.max = max.UnixNano()
+	}
+
+	// Retrieve the policy.
+	// Ignore if there are no groups created on the retention policy.
+	rp := dbi.db.policies[dbi.db.defaultRetentionPolicy]
+	if len(rp.groups) == 0 {
+		return itr
+	}
+
+	// Find all shards which match the the time range and series id.
+	// TODO: Support multiple groups.
+	g := rp.groups[0]
+
+	// Ignore shard groups that our time range does not cross.
+	if !timeBetweenInclusive(g.StartTime, min, max) &&
+		!timeBetweenInclusive(g.EndTime, min, max) {
+		return itr
+	}
+
+	// Find appropriate shard by series id.
+	sh := g.ShardBySeriesID(seriesID)
+
+	// Open a transaction on the shard.
+	tx, err := sh.store.Begin(false)
+	assert(err == nil, "read-only tx error: %s", err)
+	itr.tx = tx
+
+	// Open and position cursor.
+	b := tx.Bucket(u32tob(seriesID))
+	if b != nil {
+		cur := b.Cursor()
+		itr.k, itr.v = cur.Seek(u64tob(uint64(itr.min)))
+		itr.cur = cur
+	}
+
+	return itr
+}
+
+// iterator represents a series data iterator for a shard.
+// It can iterate over all data for a given time range for multiple series in a shard.
+type iterator struct {
+	tx       *bolt.Tx
+	cur      *bolt.Cursor
+	seriesID uint32
+	fieldID  uint8
+	typ      influxql.DataType
+
+	k, v []byte // lookahead buffer
+
+	min, max   int64 // time range
+	imin, imax int64 // interval time range
+	interval   int64 // interval duration
+}
+
+// close closes the iterator.
+func (i *iterator) Close() error {
+	if i.tx != nil {
+		return i.tx.Rollback()
+	}
+	return nil
+}
+
+// Next returns the next value from the iterator.
+func (i *iterator) Next() (key int64, value interface{}) {
+	for {
+		// Read raw key/value from lookhead buffer, if available.
+		// Otherwise read from cursor.
+		var k, v []byte
+		if i.k != nil {
+			k, v = i.k, i.v
+			i.k, i.v = nil, nil
+		} else if i.cur != nil {
+			k, v = i.cur.Next()
+		}
+
+		// Exit at the end of the cursor.
+		if k == nil {
+			return 0, nil
+		}
+
+		// Extract timestamp & field value.
+		key = int64(btou64(k))
+		value = unmarshalValue(v, i.fieldID)
+
+		// If timestamp is beyond interval time range then push onto lookahead buffer.
```

Comment:
The lookahead is necessary when we add an interval. For example, if we're doing a `GROUP BY time(1h)` then we need to read all the points up to the end of the hour. The next call to the iterator at the end of the interval will grab the point outside of the current interval, realize it's gone too far, then push it back on the buffer for the next iteration call (after we move to the next interval).